fpm/src/fpm_filesystem.f90

```diff
@@ -0,0 +1,230 @@
+    end select
+    if (stat /= 0) then
+        print *, "execute_command_line() failed"
+        error stop
+    end if
+
+end subroutine mkdir
+
+
+subroutine list_files(dir, files)
+    character(len=*), intent(in) :: dir
+    type(string_t), allocatable, intent(out) :: files(:)
+
+    integer :: stat, fh
+    character(:), allocatable :: temp_file
+
+    ! Using `inquire` / exists on directories works with gfortran, but not ifort
+    if (.not. exists(dir)) then
+        allocate(files(0))
+        return
+    end if
+
+    allocate(temp_file, source = get_temp_filename() )
+
+    select case (get_os_type())
+        case (OS_LINUX)
+            call execute_command_line("ls " // dir // " > "//temp_file, &
+                                      exitstat=stat)
+        case (OS_MACOS)
+            call execute_command_line("ls " // dir // " > "//temp_file, &
+                                      exitstat=stat)
+        case (OS_WINDOWS)
+            call execute_command_line("dir /b " // windows_path(dir) // " > "//temp_file, &
+                                      exitstat=stat)
+    end select
+    if (stat /= 0) then
+        print *, "execute_command_line() failed"
+        error stop
+    end if
+
+    open(newunit=fh, file=temp_file, status="old")
+    files = read_lines(fh)
+    close(fh,status="delete")
+
+end subroutine list_files
+
+
+logical function exists(filename) result(r)
+    character(len=*), intent(in) :: filename
+    inquire(file=filename, exist=r)
+end function
+
+
+function get_temp_filename() result(tempfile)
+    ! Get a unused temporary filename
+    !  Calls posix 'tempnam' - not recommended, but
+    !   we have no security concerns for this application
+    !   and use here is temporary.
+    ! Works with MinGW
+    !
+    use iso_c_binding, only: c_ptr, C_NULL_PTR, c_f_pointer
+    character(:), allocatable :: tempfile
+
+    type(c_ptr) :: c_tempfile_ptr
+    character(len=1), pointer :: c_tempfile(:)
+    
+    interface
+
+        function c_tempnam(dir,pfx) result(tmp) bind(c,name="tempnam")
+            import
+            type(c_ptr), intent(in), value :: dir
+            type(c_ptr), intent(in), value :: pfx
+            type(c_ptr) :: tmp
+        end function c_tempnam
+
+        subroutine c_free(ptr) BIND(C,name="free")
+            import
+            type(c_ptr), value :: ptr
+        end subroutine c_free
+
+    end interface
+
+    c_tempfile_ptr = c_tempnam(C_NULL_PTR, C_NULL_PTR)
+    call c_f_pointer(c_tempfile_ptr,c_tempfile,[LINE_BUFFER_LEN])
+
+    tempfile = f_string(c_tempfile)
+
+    call c_free(c_tempfile_ptr)
+
+end function get_temp_filename
+
+
+function windows_path(path) result(winpath)
+    ! Replace file system separators for windows
+    !
+    character(*), intent(in) :: path
+    character(:), allocatable :: winpath
+    
+    winpath = path
+
+    do while(index(winpath,'/') > 0)
+        winpath(index(winpath,'/'):index(winpath,'/')) = '\'
+    end do
```

Comment:
Probably worth using a temporary variable rather than relying on the compiler optimizing this away.